kube-controllers/pkg/controllers/loadbalancer/loadbalancer_controller.go

```diff
@@ -0,0 +1,700 @@
+			return errors.New("no IPv4 pools found from annotation")
+		}
+	}
+
+	if svc.Annotations[annotationIpv6Pools] != "" {
+		args.IPv6Pools, _ = c.resolvePools(parseAnnotation(svc.Annotations[annotationIpv6Pools]), false)
+		if args.IPv6Pools == nil {
+			return errors.New("no IPv6 pools found from annotation")
+		}
+	}
+
+	v4Assignments, v6assignments, err := c.calicoClient.IPAM().AutoAssign(c.ctx, args)
+	if err != nil {
+		log.WithField("svc", svc.Name).WithError(err).Warn("error on assigning IPAM to service")
+	}
+
+	var ingress []v1.LoadBalancerIngress
+
+	if v4Assignments != nil {
+		for _, assignment := range v4Assignments.IPs {
+			ingress = append(ingress, v1.LoadBalancerIngress{
+				IP: assignment.IP.String(),
+			})
+		}
+	}
+
+	if v6assignments != nil {
+		for _, assignment := range v6assignments.IPs {
+			ingress = append(ingress, v1.LoadBalancerIngress{
+				IP: assignment.IP.String(),
+			})
+		}
+	}
+
+	svc.Status.LoadBalancer.Ingress = ingress
+
+	_, err = c.clientSet.CoreV1().Services(svc.Namespace).UpdateStatus(c.ctx, svc, metav1.UpdateOptions{})
+	if err != nil {
+		log.Infof("LoadBalancer Error updating service %s/%s: %v", svc.Namespace, svc.Name, err)
+		return err
+	}
+	return nil
+}
+
+// releaseIP tries to release all IPs allocated with the Service unique handle
+func (c *loadBalancerController) releaseIP(handle string) error {
+	log.Info("Service type LoadBalancer deleted, releasing assigned IP address")
+
+	err := c.calicoClient.IPAM().ReleaseByHandle(c.ctx, handle)
+	if err != nil {
+		log.Errorf("error on removing assigned IP for handle %s", handle)
+		return err
+	}
+	return nil
+}
+
+// checkStatus determines what has changed about the Service. Service can change it's type, if that happens
+// we want to release the IP previously used. If calico annotations have changed we try to assign new IP address to the Service
+func (c *loadBalancerController) checkStatus(svc *v1.Service) error {
+	// Service type has changed, release the ip by the handle
+	if svc.Spec.Type != v1.ServiceTypeLoadBalancer {
+		return c.releaseIP(createHandle(svc))
+	} else {
+		// Type is load balancer, this means calico annotations have changed, and we need to re-assign IPs from this service
+		err := c.releaseIP(createHandle(svc))
+		if err != nil {
+			return err
+		}
+		// We can only assign new IP if we are still managing the address assignment
+		if isCalicoManagedLoadBalancer(svc, c.cfg.AssignIPs) {
+			return c.assignIP(svc)
+		} else {
+			// No longer managed by Calico, clean up our assigned IP from Service
+			svc.Status.LoadBalancer.Ingress = []v1.LoadBalancerIngress{}
+
+			_, err = c.clientSet.CoreV1().Services(svc.Namespace).UpdateStatus(c.ctx, svc, metav1.UpdateOptions{})
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
+
+// resolvePools valid IPpool range when specific pool is requested by the user
+func (c *loadBalancerController) resolvePools(pools []string, isv4 bool) ([]cnet.IPNet, error) {
```

Comment:
I think for clarity, we should adjust naming of variables to use a consistent scheme throughout depending on the type of object. We have a lot of things called "pools" but they can be of various types, and they are often being cross referenced with each other which makes the code harder to read:

- v3.IPPool objects
- string that represents a CIDR
- A parsed CIDR
- string that represents a pool name

I'd suggest a naming pattern like:

- pools - []v3.IPpool
- poolCIDRs - []string or []cnet.IPNet
- poolNames - []string
- poolIDs - []string (mixed name / cidr)